src/routing/routing_heuristics.jl

```diff
@@ -292,50 +292,59 @@ function optimize_waypoints(
     # Initialize variables for objective function
     wpts_length::Int64 = length(waypoints_initial) + 2 # +2 for depot start/end points
     wpts::Vector{Point{2,Float64}} = Vector{Point{2,Float64}}(undef, wpts_length)
+    wpts[1] = last_ms_route.route.nodes[1] # Depot start fixed
+    wpts[end] = last_ms_route.route.nodes[end] # Depot end fixed
+
     has_bad_waypoint::BitVector = falses(wpts_length)
     exclusion_count::Int64 = 0
     score::Float64 = 0.0
 
-    best_soln = Ref{MSTSolution}(soln)
-    best_score = Ref(Inf)
-    best_count = Ref(0)
+    best_soln::MSTSolution = soln
+    best_score::Float64 = critical_path(soln, vessel_weightings)
+    best_count::Int64 = 0
+    soln_proposed::MSTSolution = soln
 
     # Objective from x -> critical_path
     function obj(
         x::Vector{Float64};
     )::Float64
         # Rebuild waypoints
-        wpts .= Point.([
-            last_ms_route.route.nodes[1],  # first waypoint (depot)
-            tuple.(x[1:2:end], x[2:2:end])...,
-            last_ms_route.route.nodes[end]  # last waypoint (depot)
-        ])
+        @inbounds for i in 1:length(waypoints_initial)
+            wpts[i+1] = Point(x[2i-1], x[2i])
+        end
 
         # Exit early here if any waypoints are inside exclusion zones
         has_bad_waypoint = point_in_exclusion.(wpts, Ref(exclusions_mothership))
         exclusion_count = count(has_bad_waypoint)
         if exclusion_count > 0
             naive_score = sum(haversine.(wpts[1:end-1], wpts[2:end])) * vessel_weightings[1]
-            return naive_score + naive_score * exclusion_count
+            return naive_score * (exclusion_count + 1)^2
         end
 
         soln_proposed = rebuild_solution_with_waypoints(
-            best_soln[],
+            soln,
             wpts,
             exclusions_mothership,
             exclusions_tender
         )
 
         score = critical_path(soln_proposed, vessel_weightings)
+        @assert !isinf(score) "Critical path cost is infinite, indicating a bad waypoint!"
```

Comment:
Same comment as elsewhere, but also, it would be helpful for you to know what the bad waypoints were so consider including that in the error message.